spec/models/student_spec.rb

```diff
@@ -0,0 +1,326 @@
+# - CSV and YML upload
+#  - with no duplicates and no sections
+#  - with duplicates and no sections
+#  - with no duplicates and sections
+#  - with duplicates and sections
+#  - with no duplicates and one section
+#  - with duplicates and sections and update of a section
+#  - with an invalid file
+
+require 'spec_helper'
+require 'shoulda'
+
+describe Student do
+
+  # Tests for a generic Student instantiation
+  context 'A good Student model' do
+
+    it { should have_many(:accepted_groupings).through(:memberships) }
+    it { should have_many(:pending_groupings).through(:memberships) }
+    it { should have_many(:rejected_groupings).through(:memberships) }
+    it { should have_many :student_memberships }
+    it { should have_many :grace_period_deductions }
+    it { should belong_to :section }
+
+    it { should validate_numericality_of :grace_credits }
+  end
+
+  # context 'A pair of students in the same group' do
+  #     @membership1 = StudentMembership.new(membership_status: StudentMembership::STATUSES[:inviter])
+  #     @grouping = @membership1.grouping
+  #     @membership2 = StudentMembership.new({grouping: @grouping,
+  #                                            membership_status: StudentMembership::STATUSES[:accepted]})
+  #     @student1 = @membership1.user
+  #     @student2 = @membership2.user
+  #
+  #     @student_id_list = [@student1.id, @student2.id]
+  #
+  #   it 'can hide without error' do
+  #     Student.hide_students(@student_id_list)
+  #     students = Student.find(@student_id_list)
+  #
+  #     expect(students[0].hidden).to be true
+  #     expect(students[1].hidden).to be true
+  #   end
+  #
+  #   it 'should not cause error when user is not found on hide and remove' do
+  #     # Mocks to enter into the if that leads to the call to remove the student
+  #     expect(Assignment.any_instance.stubs(:vcs_submit)).to be true
+  #     expect(Grouping.any_instance.stubs(:is_valid?)).to be true
+  #
+  #     # Mock the repository and raise Repository::UserNotFound
+  #     mock_repo = mock('Repository::AbstractRepository')
+  #     expect(mock_repo.stubs(:close)).to be true
+  #     expect(mock_repo.stubs(:remove_user)).to raise_error(Repository::UserNotFound)
+  #     expect(Group.any_instance.stubs(:repo)).to eq(mock_repo)
+  #
+  #     Student.hide_students(@student_id_list)
+  #
+  #     [{type: 'negative', grace_credits: '-10', expected: 0 },
+  #     {type: 'positive', grace_credits: '10', expected: 15 }].each do |item|
+  #       it "should not error when given #{item[:type]} grace credits" do
+  #         expect(Student.give_grace_credits(@student_id_list, item[:grace_credits]))
+  #
+  #         #You have to find the students to get the updated values
+  #         students = Student.find(@student_id_list)
+  #
+  #         expect(item[:expected]).eql?(students[0].grace_credits)
+  #         expect(item[:expected]).eql?(students[1].grace_credits)
+  #       end
+  #     end
+  #   end
+  # end
```

Comment:
If code has been replaced/updated somewhere else, please remove commented out old code.  If you are saving this as a reference for later please either indicate that with a comment somewhere or add 'Work in Progress' or 'WIP' to the description of this pull request so we know you will come back to this later and the pull request should not be merged right away.